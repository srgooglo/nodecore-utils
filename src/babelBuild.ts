import babel from '@babel/core'
import yParser from 'yargs-parser'
import { join, extname, sep } from 'path'
import { existsSync, statSync, readdirSync } from 'fs'
import assert from 'assert'
import { verbosity } from './index'
import slash from 'slash2'
import rimraf from 'rimraf'
import vfs from 'vinyl-fs'
import through from 'through2'
import chokidar from 'chokidar'

const cwd = process.cwd();

let pkgCount = Number;

function getBabelConfig(isBrowser, path) {
  const targets = isBrowser
    ? {
      browsers: ['last 2 versions', 'IE 10'],
    }
    : { node: 6 };
  return {
    presets: [
      [
        require.resolve('@babel/preset-typescript'),
        {},
      ],
      [
        require.resolve('@babel/preset-env'),
        {
          targets,
          ...(isBrowser ? { modules: false } : {}),
        },
      ],
      ...(isBrowser ? [require.resolve('@babel/preset-react')] : []),
    ],
    plugins: [
      require.resolve('@babel/plugin-proposal-export-default-from'),
      require.resolve('@babel/plugin-proposal-do-expressions'),
      require.resolve('@babel/plugin-proposal-class-properties'),
    ],
  }
}

function addLastSlash(path:String) {
  return path.slice(-1) === '/' ? path : `${path}/`;
}

interface TransformTypes {
    content: any;
    path: any;
    pkg: any;
    root: any;
}

function transform(opts = {}) {
  const { content, path, pkg, root } = <TransformTypes> opts;
  assert(content, `opts.content should be supplied for transform()`);
  assert(path, `opts.path should be supplied for transform()`);
  assert(pkg, `opts.pkg should be supplied for transform()`);
  assert(root, `opts.root should be supplied for transform()`);
  assert(['.js', '.ts'].includes(extname(path)), `extname of opts.path should be .js, .ts or .tsx`);

  const { browserFiles } = pkg.umiTools || {};
  const isBrowser = browserFiles && browserFiles.includes(slash(path).replace(`${addLastSlash(slash(root))}`, ''));
  const babelConfig = getBabelConfig(isBrowser, path);
  verbosity(`transform > ${slash(path).replace(`${cwd}/`)} `)
  // @ts-expect-error
  return babel.transform(content, {
    ...babelConfig,
    filename: path,
  }).code;
}

interface BuildTypes {
    cwd: any;
    watch: any;
}

function build(dir, opts = {}) {
  const { cwd, watch } = <BuildTypes> opts;
  assert(dir.charAt(0) !== '/', `dir should be relative`);
  assert(cwd, `opts.cwd should be supplied`);

  const pkgPath = join(cwd, dir, 'package.json');
  assert(existsSync(pkgPath), 'package.json should exists');
  const pkg = require(pkgPath);
  const libDir = join(dir, 'lib');
  const srcDir = join(dir, 'src');

  // clean
  rimraf.sync(join(cwd, libDir));

  function createStream(src) {
    assert(typeof src === 'string', `src for createStream should be string`);
    return vfs
      .src([
        src,
        `!${join(srcDir, '**/fixtures/**/*')}`,
        `!${join(srcDir, '**/.umi/**/*')}`,
        `!${join(srcDir, '**/.umi-production/**/*')}`,
        `!${join(srcDir, '**/*.test.js')}`,
        `!${join(srcDir, '**/*.e2e.js')}`,
      ], {
        allowEmpty: true,
        base: srcDir,
      })
      .pipe(through.obj((f, env, cb) => {
        if (['.js', '.ts'].includes(extname(f.path)) && !f.path.includes(`${sep}templates${sep}`)) {
          f.contents = Buffer.from(
              // @ts-ignore
            transform({
              content: f.contents,
              path: f.path,
              pkg,
              root: join(cwd, dir),
            }),
          );
          f.path = f.path.replace(extname(f.path), '.js');
        }
        cb(null, f);
      }))
      .pipe(vfs.dest(libDir));
  }

  const stream = createStream(join(srcDir, '**/*'));
  stream.on('end', () => {
    // @ts-expect-error
    pkgCount -= 1;

    // @ts-expect-error
    if (pkgCount === 0 && process.send) {
      process.send('BUILD_COMPLETE');
    }
    // watch
    if (watch) {
      verbosity(`++ pending start watch > ${srcDir}`)
      const watcher = chokidar.watch(join(cwd, srcDir), {
        ignoreInitial: true,
      });
      watcher.on('all', (event, fullPath) => {
        const relPath = fullPath.replace(join(cwd, srcDir), '');
        verbosity(`** watch > [${event}] ${join(srcDir, relPath)}`)
        if (!existsSync(fullPath)) return;
        if (statSync(fullPath).isFile()) {
          createStream(fullPath);
        }
      });
    }
  });
}
// @ts-ignore
function isLerna(cwd) {
  return existsSync(join(cwd, 'lerna.json'));
}

// Init
const args = yParser(process.argv.slice(3));
const watch = args.w || args.watch;
if (isLerna(cwd)) {
  const dirs = readdirSync(join(cwd, 'packages'))
    .filter(dir => dir.charAt(0) !== '.');
        // @ts-expect-error

  pkgCount = dirs.length;
  dirs.forEach(pkg => {
    build(`./packages/${pkg}`, {
      cwd,
      watch,
    });
  });
} else {
        // @ts-expect-error

  pkgCount = 1;
  build('./', {
    cwd,
    watch,
  });
}