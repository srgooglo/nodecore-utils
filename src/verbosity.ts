
import path from 'path'
// @ts-ignore
import stackTrace from 'stack-trace'
import chalk from 'chalk'
import { __legacy__objectToArray } from '../src'

interface Obj {
    data: any,
    params: any,
    stackTraceParams: any
}

interface optStack {
    activeColor: boolean,
    line: boolean,
    method: boolean,
    file: boolean,
    time: boolean,
}

interface opt {
    color: string,
    type: string
}

export default <obj extends Readonly<Obj>>(data: obj["data"], params?: obj["params"], stackTraceParams?: obj["stackTraceParams"]) => {
    let initData;
    initData = data

    let opt = <opt>{
        color: "green",
        type: "log"
    }
    let optStackTrace = <optStack>{
        line: false,
        method: true,
        file: false,
        time: true
    }

    const frame = stackTrace.get()[1]

    const stackTraceData = {
        time: new Date().toLocaleTimeString(),
        line: `(:${frame.getLineNumber()})`,
        file: path.basename(frame.getFileName()),
        method: `[${frame.getFunctionName()}]`
    }

    if (typeof (params) !== "undefined" || params != null) {
        // @ts-ignore
        __legacy__objectToArray(params).forEach((e: any) => {
            if (typeof (e.value) !== "undefined") {
                // @ts-ignore
                opt[e.key] = e.value
            }
        })
    }

    if (typeof (stackTraceParams) !== "undefined" || stackTraceParams != null) {
        // @ts-ignore
        __legacy__objectToArray(stackTraceParams).forEach((e: any) => {
            if (typeof (e.value) !== "undefined") {
                // @ts-ignore
                optStackTrace[e.key] = e.value
            }
        })
    }

    const stackTraceKeys = Object.keys(optStackTrace)
    const stackTraceLength = stackTraceKeys.length
    let modifyCount = 0
    let tmp

    for (let i = 0; i < stackTraceLength; i++) {
        const key = stackTraceKeys[i]
        // @ts-ignore
        const value = optStackTrace[stackTraceKeys[i]]
        const divisor = (i == (stackTraceLength - 1) ? " | " : " > ")
        // console.log(`[${key}] is the ${i == stackTraceLength? "last opt" : `n[${i}]` }`)
        // console.log(i, "/", stackTraceLength -1)
        // @ts-ignore
        if (typeof (stackTraceData[key]) !== "undefined" && value) {
            if (Array.isArray(initData)) {
                if (modifyCount == 0) {
                    // @ts-ignore
                    tmp = (chalk[opt.color](`${stackTraceData[key]}`) + divisor)
                } else {
                    // @ts-ignore
                    tmp = (chalk[opt.color](`${stackTraceData[key]}`) + divisor + tmp)
                }
                if (i == (stackTraceLength - 1)) {
                    data.unshift(tmp)
                }
            } else {
                // @ts-ignore
                data = (chalk[opt.color](`${stackTraceData[key]}`) + divisor + data)
            }
            modifyCount++
        }
    }

    if (Array.isArray(data)) {
        // @ts-ignore
        return console[opt.type](...data)
    }
    // @ts-ignore
    return console[opt.type](data)
}
