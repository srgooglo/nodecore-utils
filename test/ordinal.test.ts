import { range } from 'lodash'

import { ordinal } from '../src'

const numbers = range(1, 105)

const badNumbers = [
  0,
  -2,
  3.5,
  Infinity,
]

describe('ordinal', () => {
  for (const n of numbers) {
    it(`Should ordinal(${n}) correctly`, () => {
      const result = ordinal(n)
      expect(result).toMatchSnapshot()
    })
  }

  for (const n of badNumbers) {
    it(`Should throw on ordinal(${n})`, () => {
      expect(() => ordinal(n)).toThrowErrorMatchingSnapshot()
    })
  }
})
