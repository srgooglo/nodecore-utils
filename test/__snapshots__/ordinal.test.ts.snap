// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ordinal Should ordinal(1) correctly 1`] = `"1st"`;

exports[`ordinal Should ordinal(2) correctly 1`] = `"2nd"`;

exports[`ordinal Should ordinal(3) correctly 1`] = `"3rd"`;

exports[`ordinal Should ordinal(4) correctly 1`] = `"4th"`;

exports[`ordinal Should ordinal(5) correctly 1`] = `"5th"`;

exports[`ordinal Should ordinal(6) correctly 1`] = `"6th"`;

exports[`ordinal Should ordinal(7) correctly 1`] = `"7th"`;

exports[`ordinal Should ordinal(8) correctly 1`] = `"8th"`;

exports[`ordinal Should ordinal(9) correctly 1`] = `"9th"`;

exports[`ordinal Should ordinal(10) correctly 1`] = `"10th"`;

exports[`ordinal Should ordinal(11) correctly 1`] = `"11th"`;

exports[`ordinal Should ordinal(12) correctly 1`] = `"12th"`;

exports[`ordinal Should ordinal(13) correctly 1`] = `"13th"`;

exports[`ordinal Should ordinal(14) correctly 1`] = `"14th"`;

exports[`ordinal Should ordinal(15) correctly 1`] = `"15th"`;

exports[`ordinal Should ordinal(16) correctly 1`] = `"16th"`;

exports[`ordinal Should ordinal(17) correctly 1`] = `"17th"`;

exports[`ordinal Should ordinal(18) correctly 1`] = `"18th"`;

exports[`ordinal Should ordinal(19) correctly 1`] = `"19th"`;

exports[`ordinal Should ordinal(20) correctly 1`] = `"20th"`;

exports[`ordinal Should ordinal(21) correctly 1`] = `"21st"`;

exports[`ordinal Should ordinal(22) correctly 1`] = `"22nd"`;

exports[`ordinal Should ordinal(23) correctly 1`] = `"23rd"`;

exports[`ordinal Should ordinal(24) correctly 1`] = `"24th"`;

exports[`ordinal Should ordinal(25) correctly 1`] = `"25th"`;

exports[`ordinal Should ordinal(26) correctly 1`] = `"26th"`;

exports[`ordinal Should ordinal(27) correctly 1`] = `"27th"`;

exports[`ordinal Should ordinal(28) correctly 1`] = `"28th"`;

exports[`ordinal Should ordinal(29) correctly 1`] = `"29th"`;

exports[`ordinal Should ordinal(30) correctly 1`] = `"30th"`;

exports[`ordinal Should ordinal(31) correctly 1`] = `"31st"`;

exports[`ordinal Should ordinal(32) correctly 1`] = `"32nd"`;

exports[`ordinal Should ordinal(33) correctly 1`] = `"33rd"`;

exports[`ordinal Should ordinal(34) correctly 1`] = `"34th"`;

exports[`ordinal Should ordinal(35) correctly 1`] = `"35th"`;

exports[`ordinal Should ordinal(36) correctly 1`] = `"36th"`;

exports[`ordinal Should ordinal(37) correctly 1`] = `"37th"`;

exports[`ordinal Should ordinal(38) correctly 1`] = `"38th"`;

exports[`ordinal Should ordinal(39) correctly 1`] = `"39th"`;

exports[`ordinal Should ordinal(40) correctly 1`] = `"40th"`;

exports[`ordinal Should ordinal(41) correctly 1`] = `"41st"`;

exports[`ordinal Should ordinal(42) correctly 1`] = `"42nd"`;

exports[`ordinal Should ordinal(43) correctly 1`] = `"43rd"`;

exports[`ordinal Should ordinal(44) correctly 1`] = `"44th"`;

exports[`ordinal Should ordinal(45) correctly 1`] = `"45th"`;

exports[`ordinal Should ordinal(46) correctly 1`] = `"46th"`;

exports[`ordinal Should ordinal(47) correctly 1`] = `"47th"`;

exports[`ordinal Should ordinal(48) correctly 1`] = `"48th"`;

exports[`ordinal Should ordinal(49) correctly 1`] = `"49th"`;

exports[`ordinal Should ordinal(50) correctly 1`] = `"50th"`;

exports[`ordinal Should ordinal(51) correctly 1`] = `"51st"`;

exports[`ordinal Should ordinal(52) correctly 1`] = `"52nd"`;

exports[`ordinal Should ordinal(53) correctly 1`] = `"53rd"`;

exports[`ordinal Should ordinal(54) correctly 1`] = `"54th"`;

exports[`ordinal Should ordinal(55) correctly 1`] = `"55th"`;

exports[`ordinal Should ordinal(56) correctly 1`] = `"56th"`;

exports[`ordinal Should ordinal(57) correctly 1`] = `"57th"`;

exports[`ordinal Should ordinal(58) correctly 1`] = `"58th"`;

exports[`ordinal Should ordinal(59) correctly 1`] = `"59th"`;

exports[`ordinal Should ordinal(60) correctly 1`] = `"60th"`;

exports[`ordinal Should ordinal(61) correctly 1`] = `"61st"`;

exports[`ordinal Should ordinal(62) correctly 1`] = `"62nd"`;

exports[`ordinal Should ordinal(63) correctly 1`] = `"63rd"`;

exports[`ordinal Should ordinal(64) correctly 1`] = `"64th"`;

exports[`ordinal Should ordinal(65) correctly 1`] = `"65th"`;

exports[`ordinal Should ordinal(66) correctly 1`] = `"66th"`;

exports[`ordinal Should ordinal(67) correctly 1`] = `"67th"`;

exports[`ordinal Should ordinal(68) correctly 1`] = `"68th"`;

exports[`ordinal Should ordinal(69) correctly 1`] = `"69th"`;

exports[`ordinal Should ordinal(70) correctly 1`] = `"70th"`;

exports[`ordinal Should ordinal(71) correctly 1`] = `"71st"`;

exports[`ordinal Should ordinal(72) correctly 1`] = `"72nd"`;

exports[`ordinal Should ordinal(73) correctly 1`] = `"73rd"`;

exports[`ordinal Should ordinal(74) correctly 1`] = `"74th"`;

exports[`ordinal Should ordinal(75) correctly 1`] = `"75th"`;

exports[`ordinal Should ordinal(76) correctly 1`] = `"76th"`;

exports[`ordinal Should ordinal(77) correctly 1`] = `"77th"`;

exports[`ordinal Should ordinal(78) correctly 1`] = `"78th"`;

exports[`ordinal Should ordinal(79) correctly 1`] = `"79th"`;

exports[`ordinal Should ordinal(80) correctly 1`] = `"80th"`;

exports[`ordinal Should ordinal(81) correctly 1`] = `"81st"`;

exports[`ordinal Should ordinal(82) correctly 1`] = `"82nd"`;

exports[`ordinal Should ordinal(83) correctly 1`] = `"83rd"`;

exports[`ordinal Should ordinal(84) correctly 1`] = `"84th"`;

exports[`ordinal Should ordinal(85) correctly 1`] = `"85th"`;

exports[`ordinal Should ordinal(86) correctly 1`] = `"86th"`;

exports[`ordinal Should ordinal(87) correctly 1`] = `"87th"`;

exports[`ordinal Should ordinal(88) correctly 1`] = `"88th"`;

exports[`ordinal Should ordinal(89) correctly 1`] = `"89th"`;

exports[`ordinal Should ordinal(90) correctly 1`] = `"90th"`;

exports[`ordinal Should ordinal(91) correctly 1`] = `"91st"`;

exports[`ordinal Should ordinal(92) correctly 1`] = `"92nd"`;

exports[`ordinal Should ordinal(93) correctly 1`] = `"93rd"`;

exports[`ordinal Should ordinal(94) correctly 1`] = `"94th"`;

exports[`ordinal Should ordinal(95) correctly 1`] = `"95th"`;

exports[`ordinal Should ordinal(96) correctly 1`] = `"96th"`;

exports[`ordinal Should ordinal(97) correctly 1`] = `"97th"`;

exports[`ordinal Should ordinal(98) correctly 1`] = `"98th"`;

exports[`ordinal Should ordinal(99) correctly 1`] = `"99th"`;

exports[`ordinal Should ordinal(100) correctly 1`] = `"100th"`;

exports[`ordinal Should ordinal(101) correctly 1`] = `"101st"`;

exports[`ordinal Should ordinal(102) correctly 1`] = `"102nd"`;

exports[`ordinal Should ordinal(103) correctly 1`] = `"103rd"`;

exports[`ordinal Should ordinal(104) correctly 1`] = `"104th"`;

exports[`ordinal Should throw on ordinal(-2) 1`] = `"Number must be a positive integer. Instead got: -2"`;

exports[`ordinal Should throw on ordinal(0) 1`] = `"Number must be a positive integer. Instead got: 0"`;

exports[`ordinal Should throw on ordinal(3.5) 1`] = `"Number must be a positive integer. Instead got: 3.5"`;

exports[`ordinal Should throw on ordinal(Infinity) 1`] = `"Number must be a positive integer. Instead got: Infinity"`;
